locks and docs in ATMP and CheckInputsFromMempoolAndCache

glozow opened the pull request:
This is a very small PR that adds some lock annotations to clarify that, now, the `pool.cs` lock is held throughout tx validation for mempool.  The comments in `CheckInputsFromMempoolAndCache` were unclear/outdated so I updated those as well.

~This PR is a cleanup. It removes unnecessary code that doesn't do much.~

glozow committed:
lock annotations for MemPoolAccept functions

We should already have the mempool lock when entering
CheckInputsFromMempoolAndCache

jnewbery reviewed (commented):
Concept ACK.

I think you could update the PR description to say that CIFMAC also does:

- `CheckInputScripts() to cache signature and script validity against current tip consensus rules.`

in order for script caching to work, and that you're retaining that check and moving it into the calling function (`ConsensusScriptChecks()`).

michaelfolkson commented:
Concept ACK. Great PR description, really informative. 

Is there anything in particular you'd recommend a reviewer do to assure themselves that removing this code doesn't cause any problems beyond the usual building, running tests etc?

MarcoFalke commented on src/validation.cpp:
```diff
@@ -404,43 +404,6 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
-        } else {
-            const Coin& coinFromDisk = ::ChainstateActive().CoinsTip().AccessCoin(txin.prevout);
-            assert(!coinFromDisk.IsSpent());
-            assert(coinFromDisk.out == coin.out);
```
I assumed this was protecting against corruption of the utxo set on disk, but that doesn't seem to be the case. `coin` will be from disk as well.

jnewbery replied:
Is that true? `AccessCoin()` doesn't necessarily hit the disk. It could be pulling the coin from the cache.

MarcoFalke replied:
in which case `coin` would also be from the cache, no?

maflcko reviewed (commented):
Approach ACK. The code indeed doesn't seem to do anything.

glozow commented:
@michaelfolkson good question. You could go line-by-line in the function and verify that what I said in the description is true, i.e. that each one is done earlier in ATMP. And then you can also look at what `cs_main` and `pool.cs` guarantee wrt consistency so you can say "yeah, this wouldn't change." Might be fun to trace the coinsview through ATMP with some logs. If you prefer writing code, you can try to break it, e.g. with a [functional test](https://github.com/bitcoin/bitcoin/pull/20833/files#diff-2757e99ad43064441a6401f5002232a6a311d92d16b49df66f51d427ce448a68R86) with badly formed/similar but conflicting Coins.

jnewbery commented:
This code was added in commit https://github.com/bitcoin/bitcoin/commit/b014668e27b496bd6ad30985294f3d6971311910, specifically in response to this review comment: https://github.com/bitcoin/bitcoin/pull/10192#issuecomment-295316813.

The code was added here: https://github.com/bitcoin/bitcoin/pull/10192#event-1049081888 before the change to per-txout db and later rebased: https://github.com/bitcoin/bitcoin/pull/10192#issuecomment-306359178.

I believe the reasoning is as follows:

- if there's a bug in CCoinsViewMempool (e.g. returning the wrong coin when looking up by txid), then the script cache could cache an incorrect outcome for script validation. That is then used as an input to validating a block, meaning that CCoinsViewMempool has become part of consensus.
- therefore, explicitly check that the hash of scriptPubKey returned by CCoinsViewMempool for this outpoint is correct. Either:
  - fetch the entire transaction from the mempool via `pool.get()` and check that the hash is committing to the scriptPubKey; or
  - lookup the coin in the UTXO set (pcoinsTip) and check that the scriptPubKey matches. Despite the name `coinFromDisk`, I believe that the coin could be fetched either from disk or the in-memory pcoinsTip cache. pcoinsTip is already part of consensus, so if it's returning bad data, then we're already in trouble.

I'm not sure we really need this. We could move these checks to immediately after the coins are fetched from CCoinsViewMempool. At that point we're basically adding duplicate code to ensure that there isn't a bug in CCoinsViewMempool, which itself is only a few lines of code. We may as well just move CCoinsViewMempool into validation.cpp as suggested in the original PR by @sdaftuar:

> as per offline discussion, perhaps just move CCoinsViewMemPool into validation.cpp to make it clear it's part of consensus, and sanity check the results from the mempool.

Another interesting thing to note is that at the time this was introduced, the mempool lock was not held for the duration of ATMP (e.g. it'd be released after this block: https://github.com/bitcoin/bitcoin/blob/b014668e27b496bd6ad30985294f3d6971311910/src/validation.cpp#L528-L572). That was added later in #12368. I don't think it actually changes anything for the purposes of ATMP (since cs_main was held for the duration), but it does make expectations clearer.

sdaftuar commented:
I'm pretty skeptical of this change.  I think the OP downplays the consensus risk if a bug is introduced in the CCoinsViewMempool or the mempool (which CCoinsViewMempool relies on).

I also took a quick look at #20833, and it seems to make that code much higher risk if we would be turning CCoinsViewMempool into something consensus critical and then make a bunch of changes to it.  Is there a method for implementing #20833 that does not require this?

(I was going to suggest looking at what I proposed in #16401, which faced the same issue, but now I suspect there may be a consensus bug there as well -- yikes.)

jnewbery commented on src/validation.cpp:
```diff
@@ -167,6 +167,24 @@ namespace {
+
+bool CCoinsViewMemPool::GetCoin(const COutPoint &outpoint, Coin &coin) const {
+    // If an entry in the mempool exists, always return that one, as it's guaranteed to never
+    // conflict with the underlying cache, and it cannot have pruned entries (as it contains full)
+    // transactions. First checking the underlying cache risks returning a pruned entry instead.
```
Everything after ", and it cannot have pruned ..." can be removed. There's no such thing as pruned coins cache entries since per-txout db.

glozow replied:
Oh true. I do fix this in #20833... should I do it here instead?

jnewbery replied:
Up to you. I only mention it here because you're touching this code, but if you want to keep this move-only, then that's fine as well.

glozow commented:
@jnewbery @sdaftuar Thanks for your input and for digging into things! I agree that a bug here would be awful as we could potentially cache incorrect script checks, affecting consensus. I think these lines from CIFMAC are a nice sanity check:
``` c
        const CTransactionRef& txFrom = m_pool.get(txin.prevout.hash);
        if (txFrom) {
            assert(txFrom->GetHash() == txin.prevout.hash);
            assert(txFrom->vout.size() > txin.prevout.n);
            assert(txFrom->vout[txin.prevout.n] == coin.out);
        } else {
            const Coin& coinFromDisk = ::ChainstateActive().CoinsTip().AccessCoin(txin.prevout);
            assert(!coinFromDisk.IsSpent());
            assert(coinFromDisk.out == coin.out);
        }
```

The main issues are with them being so late in ATMP.
> > as per offline discussion, perhaps just move CCoinsViewMemPool into validation.cpp to make it clear it's part of consensus, and sanity check the results from the mempool.

> We could move these checks to immediately after the coins are fetched from CCoinsViewMempool.

I've moved `CCoinsViewMemPool` to validation and moved the sanity checks to `PreChecks` (instead of removing them), right after we get the coins. Now that we hold the lock throughout ATMP, I hope it's clear we don't need to do these exact same checks again later.
> I believe #16401 had a bug relating to packages of transactions that conflict with themselves, because the duplicate inputs check only occurs on a per-transaction basis, so when processing a package we need to check for duplicate inputs across the whole package. But that seems like an easy fix (and #20833 seems to address this exact issue already)

Yes exactly, this cleanup came up while I was trying to figure out conflicting transactions. I did a light rebase of #20833 on top of this and [this condition](https://github.com/glozow/bitcoin/commit/df88ce2761d3b9ff16949864c3f710b3ad5761ee#) needs to be added, but otherwise it still works.

sheffine commented:
I am not sure if this message is for me ,However I wont touch anything
until you return a pruned entry.


On Tue, Jan 5, 2021 at 3:10 PM John Newbery <notifications@github.com>
wrote:

> *@jnewbery* commented on this pull request.
> ------------------------------
>
> In src/validation.cpp
> <https://github.com/bitcoin/bitcoin/pull/20834#discussion_r552166146>:
>
> > +    // conflict with the underlying cache, and it cannot have pruned entries (as it contains full)
> +    // transactions. First checking the underlying cache risks returning a pruned entry instead.
>
> Up to you. I only mention it here because you're touching this code, but
> if you want to keep this move-only, then that's fine as well.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/20834#discussion_r552166146>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ3ZHMOQG7YAFUBPVIQQI5TSYNW2BANCNFSM4VRU5IHQ>
> .
>

jnewbery commented:
I like the way this is going. CCoinsViewMempool should live in validation.cpp and be considered consensus critical since script caching was introduced. Making that clear by reflecting it in the code structure, rather than working around it by double checking the results returned from CCoinsViewCache seems like an improvement.

As far as I'm aware, CTxMemPool is not consensus critical, since any bugs in there would not cause us to incorrectly validate a block. Making the boundary between consensus-critical and non-consensus-critical code clearer is a good goal and fits with the current direction of the project (eg see #20158).

It'd be nice to eventually have ChainstateActive expose a GetCoin() method that returns a coin from the UTXO set or mempool so that callers (rpc, rest, etc) don't need to concern themselves with locking cs_main/mempool.cs and setting the coins view. After that happens, CCoinsViewMempool can exist just in validation.cpp and not be exposed in the header.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -658,6 +639,19 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
             // Otherwise assume this might be an orphan tx for which we just haven't seen parents yet
             return state.Invalid(TxValidationResult::TX_MISSING_INPUTS, "bad-txns-inputs-missingorspent");
         }
+        // Sanity check each input.
```
nit: Could explain a bit more what this sanity check is doing? Maybe a oneline summary of https://github.com/bitcoin/bitcoin/pull/20834#issuecomment-754652998 ?

glozow replied:
Made big effort to improve the documentation :D

sdaftuar commented:
Concept NACK.  I think the current structure of using CIFMAC to encapsulate the consensus requirements of the script cache is better than spreading the checks out across validation, where it is less clear why these checks exist and what they protect against.  Moreover, I don't believe this change is necessary (or desired) for #20833, as I commented here: https://github.com/bitcoin/bitcoin/pull/20833#discussion_r552684985

glozow committed:
[doc] for CheckInputsFromMempoolAndCache

glozow commented:
@sdaftuar Thanks for your thorough review - I'm no longer removing CIFMAC. I think the lock annotations are still helpful, though, and in general would like to update the documentation to clarify what this function does.

MarcoFalke commented:
Concept ACK for updating locks and docs

jnewbery reviewed (commented):
utACK 89da5ed578b4839c426151960594bd4a9bf6d4f7

This is a definite improvement to the documentation/naming.

It's not clear to me why we want `CheckInputsFromMempoolAndCache()` to be a separate function from `ConsensusScriptChecks()`. It means that the documentation is split out over three places - the `ConsensusScriptChecks()` declaration ("Re-run the script checks, using consensus flags, and try to cache the result in the scriptcache..."), the `ConsensusScriptChecks()` function body ("Check again against the current block tip's script verification flags to cache our script execution flags. This is, of course, useless...") and the `CheckInputsFromMempoolAndCache()` function comment ("Checks to avoid mempool polluting consensus critical paths ..."). It would be much clearer if the code from `CheckInputsFromMempoolAndCache()` was brought into `ConsensusScriptChecks()` and fully documented in one place.

Maybe there's a reason that the code is structured this way and the intent is to call `CheckInputsFromMempoolAndCache()` from more than one place. If not, then I think a future PR could consolidate them. In any case, this PR is an improvement and stands on its own.

sdaftuar commented:
> It's not clear to me why we want CheckInputsFromMempoolAndCache() to be a separate function from ConsensusScriptChecks(). It means that the documentation is split out over three places - the ConsensusScriptChecks() declaration ("Re-run the script checks, using consensus flags, and try to cache the result in the scriptcache..."), the ConsensusScriptChecks() function body ("Check again against the current block tip's script verification flags to cache our script execution flags. This is, of course, useless...") and the CheckInputsFromMempoolAndCache() function comment ("Checks to avoid mempool polluting consensus critical paths ..."). It would be much clearer if the code from CheckInputsFromMempoolAndCache() was brought into ConsensusScriptChecks() and fully documented in one place.

Note that `ConsensusScriptChecks` really exists for two reasons: it's a safeguard against policy script checks somehow not being a superset of consensus script checks (a software bug that could lead to invalid blocks being produced), and it conveniently hooks into caching script validation success to serve as a huge optimization for block validation.  If this is unclear, then the documentation should be improved.

> Maybe there's a reason that the code is structured this way and the intent is to call CheckInputsFromMempoolAndCache() from more than one place. If not, then I think a future PR could consolidate them. In any case, this PR is an improvement and stands on its own.

Somewhere we should draw a line between mempool validation and block validation; our philosophy has long been to not allow mistakes during mempool validation to cause block validation errors, because we have had historical bugs that have cropped up over the years that could have either caused consensus splits or invalid blocks to be mined. (One example that I recall is #6077, which could have broken consensus;  I heard stories that prior to my time working on this project, there were prior bugs that allowed invalid transactions in the mempool, which is the motivation for `TestBlockValidity` to be called before returning block candidates to miners.)

Script caching of course is a huge performance win, justifying a desire to have mempool validation interact with block validation.  So an important part of how we use mempool validation to speed up block validation is designing a system that is robust to differences in mempool vs block validation (the script cache design itself tries to address many of these concerns, in comparison with prior attempts to cache validation success).  And in particular, finding a way to draw the line between mempool and block validation so that any future bugs introduced in mempool validation would not result in a consensus split is also very important.

Drawing this line at CIFMAC seems to have held up pretty well.  At the least, mempool logic was refactored in preparation for package validation back in #16400 (which is when `MemPoolAccept::ConsensusScriptChecks` was introduced, after CIFMAC's introduction) and is being modified again a bit in #20833, yet CIFMAC doesn't need to be touched to accommodate the feature (and indeed seems mostly untouched since first introduced in 2017).  I think that is a positive sign that we've drawn a reasonably good boundary, as it seems to allow for improvements in one side of the code without touching consensus.

So while we could move this consensus critical function into the `MemPoolAccept::ConsensusScriptChecks()` function and eliminate a function call, I think doing so would blur the line rather than clarify it (for instance, CIFMAC doesn't need access to the `MemPoolAccept` state), and would make it more likely that we introduce consensus breaking bugs around the script cache in the future when code relating to mempool acceptance is modified again.  At the least, it would raise the review burden for any future PR that would touch mempool acceptance / `ConsensusScriptChecks` and which otherwise wouldn't need to touch CIFMAC, in its current form.  I think this would be a bad idea.

At a higher level: I really think that we should have a very high bar for refactoring or otherwise changing consensus-critical code -- which CIFMAC most certainly is (as the interface to the script cache), and this PR didn't seem to acknowledge when first opened.  Sometimes we'll have to modify consensus critical code to fix a bug or implement a feature, and we'll take that review burden and effort when necessary, but when unnecessary I think the cost/benefit of changing consensus code tilts strongly towards not making changes.

As for the current version of this PR: @glozow thank you for taking the feedback!  At first glance it looks much better to me as just doc changes and lock annotations -- will take a closer look.

sdaftuar reviewed (commented):
Concept ACK -- just a couple small questions.

glozow commented on src/validation.cpp:
```diff
@@ -404,36 +404,35 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
+// transaction again during block validation.
 static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, TxValidationState& state, const CCoinsViewCache& view, const CTxMemPool& pool,
-                 unsigned int flags, PrecomputedTransactionData& txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+                 unsigned int flags, PrecomputedTransactionData& txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main, pool.cs) {
```
Done!

glozow commented on src/validation.cpp:
```diff
@@ -407,22 +407,17 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
-        if (coin.IsSpent()) return false;
+        // This coin was checked in PreChecks and MemPoolAccept
+        // has been holding cs_main since then.
+        assert(!coin.IsSpent());
```
Ah, I wasn't aware of that! Yes I think that's better.

glozow commented on src/validation.cpp:
```diff
@@ -502,13 +497,13 @@ class MemPoolAccept
     // Run the script checks using our policy flags. As this can be slow, we should
     // only invoke this on transactions that have otherwise passed policy checks.
-    bool PolicyScriptChecks(ATMPArgs& args, const Workspace& ws, PrecomputedTransactionData& txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    bool PolicyScriptChecks(ATMPArgs& args, const Workspace& ws, PrecomputedTransactionData& txdata) EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs);
```
I've added annotations to `CheckFeeRate` for now.

In my mind, I think of `MemPoolAccept` class as one "validation session" on top of current Tip and mempool, so would want to assert that it holds `cs_main, pool.cs` for the entirety of `AcceptSingleTransaction`. Don't know how to do that though 😅 not familiar

TheBlueMatt replied:
The non-static functions need EXCLUSIVE_LOCKS_REQUIRED in any header files which include them as well - EXCLUSIVE_LOCKS_REQUIRED will be ignored by call-sites if its only set in the definitions and not all declarations.

sdaftuar commented on src/validation.cpp:
```diff
@@ -407,22 +407,17 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
-        if (coin.IsSpent()) return false;
+        // This coin was checked in PreChecks and MemPoolAccept
+        // has been holding cs_main since then.
+        Assume(!coin.IsSpent());
```
I believe we need to leave in the old line that returned false if this test fails (after this `Assume()` call); this check is a no-op in non-debug builds, so we'd want to return false if the check were to fail.

glozow replied:
oops! fixed

sdaftuar commented:
utACK 2f463f57e3a9797236142a525703359a98fe19ea

maflcko reviewed (approved):
cr ACK 2f463f57e3a9797236142a525703359a98fe19ea

glozow commented on src/validation.cpp:
```diff
@@ -407,21 +407,17 @@ static void UpdateMempoolForReorg(CTxMemPool& mempool, DisconnectedBlockTransact
-        // here then return false.
+        // This coin was checked in PreChecks and MemPoolAccept
+        // has been holding cs_main since then.
+        Assume(!coin.IsSpent());
```
I had `Assert` and changed to `Assume` based on @sdaftuar's [comment](https://github.com/bitcoin/bitcoin/pull/20834#discussion_r554095702) 😅 

> we don't need to crash if this assumption is violated (as the old code handled this case and returned false, so we could just continue to do that) -- but adding the Assume() will mean we do get the crash in debug builds, which I think is what we're looking for.

MarcoFalke replied:
Just a nit. It seems a bit odd to assume the `Assert` here could fail and should be an `Assume` instead. Then, at the same time assume that the other asserts (before and after this one) checking the exact same thing won't hit. Seems more consistent to have all three be `Assert`, or even remove this one completely and rely on the others.

sdaftuar replied:
If we thought the Assume could fail, we would fix the code, not add an Assume.  I think it's a question of what do you want the software to do if we are all somehow wrong in our reasoning.  In my view, crashing -- which could be a network-wide event -- should be avoided if the software is otherwise able to recover, and since that seems to be the case here, I don't know why we would change the code to crash.  It just makes this change needlessly risky, even if it's low-probability.

I think that there are other uses of assert in our codebase that I'd probably disagree with too, which is why I don't find it convincing if the argument is that some other assert would already trigger.

I agree that this is a nit either way and if it had already been written with an assert, I wouldn't be likely to open a PR suggesting that we change it. But I would personally prefer to avoid ACKing a change that is later found to cause a network-wide crash when fed the wrong data over the wire, simply because of an oversight during the review process!

MarcoFalke replied:
Thanks, makes sense

jnewbery commented:
utACK 2f463f57e3a9797236142a525703359a98fe19ea

fanquake reviewed (approved):
ACK 2f463f57e3a9797236142a525703359a98fe19ea
